src/ffi.rs

```diff
@@ -92,15 +92,15 @@ pub fn balance(args: i32, len: i32) -> i64 {
         Err(_) => return utils::fail(),
     };
 
-    let mut keys = unsafe { [mem::zeroed(); MAX_KEY + 1] };
+    let mut keys = unsafe { [mem::zeroed(); MAX_KEY] };
     let mut values = Vec::with_capacity(notes.len());
     let mut keys_len = 0;
     let mut sum = 0u64;
 
     'outer: for note in notes {
         // we iterate all the available keys until one can successfully decrypt
         // the note. if all fails, returns false
-        for idx in 0..=MAX_KEY {
+        for idx in 0..MAX_KEY {
             if keys_len == idx {
                 keys[idx] = key::derive_vk(&seed, idx as u64);
```

Comment:
Also here, you shouldn't `derive` the keys in the loop, for the same reason we discussed above. You should just instatiate outside as we did above.